processor/agentmetricsprocessor/utils_calculate_utilizations.go

```diff
@@ -0,0 +1,263 @@
+// a usage metric across one label (dimension) using the formula:
+//
+// value{l1=v1,...} = value{l1=v1,...} / sum(value{l1=vx,...}) over x=1..N
+
+const (
+	cpuTime         = "system.cpu.time"
+	memoryUsage     = "system.memory.usage"
+	fileSystemUsage = "system.filesystem.usage"
+	swapUsage       = "system.swap.usage"
+)
+
+var metricsToComputeUtilizationFor = map[string]bool{
+	cpuTime:         true,
+	memoryUsage:     true,
+	fileSystemUsage: true,
+	swapUsage:       true,
+}
+
+const stateLabel = "state"
+
+func (mtp *agentMetricsProcessor) appendUtilizationMetrics(rms pdata.ResourceMetricsSlice) error {
+	for i := 0; i < rms.Len(); i++ {
+		ilms := rms.At(i).InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			metrics := ilms.At(j).Metrics()
+			for k := 0; k < metrics.Len(); k++ {
+				metric := metrics.At(k)
+
+				// ignore all metrics except the ones we want to compute utilizations for
+				metricName := metric.MetricDescriptor().Name()
+				if _, ok := metricsToComputeUtilizationFor[metricName]; !ok {
+					continue
+				}
+
+				// calculate new utilization metric and append it
+				utilizationMetric, err := mtp.calculateUtilizationMetric(metric)
+				if err != nil {
+					return err
+				}
+
+				metrics.Append(&utilizationMetric)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (mtp *agentMetricsProcessor) calculateUtilizationMetric(metric pdata.Metric) (pdata.Metric, error) {
+	utilizationMetric := pdata.NewMetric()
+	metric.CopyTo(utilizationMetric)
+	utilizationMetric.MetricDescriptor().SetName(metricPostfixRegex.ReplaceAllString(metric.MetricDescriptor().Name(), "utilization"))
+	utilizationMetric.MetricDescriptor().SetType(pdata.MetricTypeDouble)
+
+	// for "cpu.time", we need to convert cumulative values to delta values before
+	// computing utilization of the deltas
+	isCPUTime := metric.MetricDescriptor().Name() == cpuTime
+	if isCPUTime {
+		mtp.convertPrevCPUTimeToDelta(utilizationMetric)
+	}
+
+	if err := calculateUtilizationFromDoubleDataPoints(utilizationMetric); err != nil {
+		return pdata.NewMetric(), err
+	}
+
+	if err := calculateUtilizationFromInt64DataPoints(utilizationMetric); err != nil {
+		return pdata.NewMetric(), err
+	}
+
+	// persist the values of "cpu.time" so we can compute deltas on the next cycle
+	if isCPUTime {
+		mtp.prevCPUTimeValues = doubleDataPointsToMap(metric)
+	}
+
+	return utilizationMetric, nil
+}
+
+// convertPrevCPUTimeToDelta converts the cpu.time values to delta values using the
+// values persisted in the previous cycle
+func (mtp *agentMetricsProcessor) convertPrevCPUTimeToDelta(metric pdata.Metric) {
+	ddps := metric.DoubleDataPoints()
+	for i := 0; i < ddps.Len(); {
+		ddp := ddps.At(i)
+
+		// if we have no previous value for this cpu/state combination,
+		// remove the data point as we cannot calculate a utilization
+		prevValue, ok := mtp.prevCPUTimeValues[labelsAsKey(ddp.LabelsMap())]
+		if !ok {
+			ddps.At(ddps.Len() - 1).CopyTo(ddp)
```

Comment:
So you're overwriting the current element with the value from the last element, and then remove the last element.
The order of the data points may change, but that's fine?

Reply:
What if `i == ddps.Len() - 1`? Won't we try to rewrite the element with itself in that case?
(This is probably a no-op, so may work fine).

Reply:
Yup, within a given snapshot, we don't care about the order of the data points.

Yes you're right if `i == ddps.Len() - 1` we would overwrite the element with itself and then remove it. Moved this into a `removeElementAt` function so its a little clearer